content/posts/2021/07-15-salt-nornir-intro.md

```diff
@@ -0,0 +1,229 @@
+                      +------------+--------------+               
+                                   |                     
+                    |           MASTER             |
+                    |                              |
+                    +----------Server/VM-----------+
+    ENGINES                        |                       API
+       |                           |                        |
+       |                           |                        |
+===========================ZeroMQ  EVENT BUS=============================
+                 |                                    |
+                 |                                    |
+      +------Server/VM-------+              +------Server/VM-------+
+      |                      |              |                      |
+      |  OS (Windows/Linux)  |              |  OS (Windows/Linux)  |
+      |   |-MINION process   |              |   |-PROXY-MINION     |   
+      |     |-BEACON         |              |     |-BEACON         |   
+      |     |-GRAINS         |              |     |-GRAINS         |   
+      |     |-RETURNER       |              |     |-RETURNER       |   
+      |     |-SCHEDULER      |              |     |-SCHEDULER      |   
+      |     |-EXECUTION      |              |     |-EXECUTION      |
+      |       MODULES        |              |       MODULES        |
+      |                      |              |                      |
+      +----------------------+              +----------------------+
+                                                      |         
+                                                   SSH/API    
+                                                      |         
+                                                   +------+     
+                                                   |DEVICE|     
+                                                   +------+ 
+```
+SALT Master is the HUB of the overall system, it communicates with Minions to execute various tasks. SALTSTACK uses [modules](https://docs.saltproject.io/en/latest/ref/index.html) to address various use cases.
+
+In the simplest (master-less) case, no Master required and only Minion process need to run on the system. But (probably) most common deployment is a two tier hierarchy where Master controls (many) Minions. Each Minion acts as an agent natively running on the Operating System of the system being managed. 
+
+Two-tier approach works well while you need to manage devices that can run Python and other packages required by Minion. But it falls short when you need to manage systems that does not provide such capabilities and instead could be managed over API (HTTP, Netconf etc.) or SSH/Telnet.
+
+Three-tier hierarchy was developed to accommodate for systems that cannot run Minion processes. For that case, special minion process need to run somewhere where it is reachable by Master and able to talk with managed device - this type of minion called proxy-minion. 
```

Comment:
"A three-tier"

"to accommodate systems" (remove "for")

"by THE Master"

"with THE managed device"

"IS called proxy-minion"